### PR TITLE
made RepeatedField<T> throw InvalidOperationException when collection…

### PR DESCRIPTION
**Pull request by andreabalducci**
… changes during iteration

### CODE REVIEW

**Comment by andreabalducci** [comment]
I signed it!

**Comment by xfxyjwf** [comment]
@anandolee

File: csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs [hunk 1]
```diff
@@ -303,6 +303,19 @@ public void Enumerator()
             }
         }
 
+        [Test]
+        public void Throw_InvalidOperation_Exception_if_modified()
+        {
+            var list = new RepeatedField<string> { "first", "second" };
+            using (var enumerator = list.GetEnumerator())
+            {
+                list.RemoveAt(0);
```
Comment (anandolee) [inline csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: What if remove one item and then add one item back during enumeration operation?

Comment (andreabalducci) [reply csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: it's possible but unlikely. I just covered the most common scenario to limit the footprint of the patch. 

A better approach would be to clone the internal array and enumerate the snapshot but will impact memory usage. 80% of the time you'll enumerate to consume repeated field.

An even safer approach would be adding checks on array mutations while an enumeration is in progress.

Comment (anandolee) [reply csharp/src/Google.Protobuf.Test/Collections/RepeatedFieldTest.cs]: Does "add checks on array mutations while enumeration is in progress" cost lot? If no can you change to this way?

**Comment by anandolee** [comment]
Friendly ping, andreabalducci@ are you still working on this PR ?

**Comment by anandolee** [comment]
I am closing it for no response about 4 months. Feel free to reopen if still like to go with the PR

**Pull request closed without merging**